### PR TITLE
Support YANG schema simplification, data parser generation

### PR DESCRIPTION
**Pull request by wingo**
Right now we have a YANG schema parser in `lib.yang.schema`, which uses a YANG syntax parser from `lib.yang.parser`.  That schema support is augmented by support for YANG data, implemented in `lib.yang.yang` and `lib.yang.helpers`.

Since we designed YANG support, some requirements have become clearer regarding Snabb program configuration, and some have changed.  In particular in the future I think we will see parsing and validation happening only in the world-to-snabb-config-process direction, and that the config process will hand a binary data file to the dataplane, which it will load without looking at the YANG schema, and which it will assume is correct.  Queries from the world to the config process will not validate the data, because we assume that the data plane, both config and state data, is adequately modelled by the YANG schema.

For that reason the YANG data tree doesn't need to be a tree of boxes with getters and setters that do validation.

Furthermore, one thing that I think we missed earlier was the need to formalize the mapping between a YANG schema and a grammar for device configuration (#519) and to define a grammar for a binary file (#517).
#519 implies the need to simplify a schema as well, lowering it to a kind of "primitive YANG" in which there are no "grouping" and "uses" statements because all "uses" have been replaced by the appropriate "grouping", and similarly for module imports, etc etc.

This simplification pass is really a source-to-source transformation over a programming language term; like you can think of a "grouping" statement as declaring a procedure, and a "uses" statement as calling that procedure, and simplification is beta reduction (inlining).  So, our passes should look more like compiler passes.  I prefer a functional style when writing compilers -- you can mutate things locally but calling a procedure generally should not mutate its arguments.  Otherwise I get all confused and I can't reason well about what the compiler does.

That last paragraph means that we shouldn't make `local feature = Feature.new(parent)` add the resulting feature to the parent -- we should have a less OO style and a more functional style IMO.  I am verrrrry ready to admit ignorance in many domains but for compilers I do have a lot of experience and if you are willing to assent @xray7224 then it would be a delight to be on the same page :)

The points above add up to a need to rewrite the YANG facilities, I think, and this PR is an effort in that direction.  The remaining items to do are to rename `schema2.lua` to `schema.lua`, to add some documentation, and replace the `yang.lua` module.  What do you think @xray7224 ?

### CODE REVIEW

**Comment by wingo** [comment]
Another missing thing is primitive support for IPv4 / IPv6 / etc addresses.  We will still want to treat these as primitive types I think, as we do now.

File: src/lib/yang/data.lua [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+local function integer_type(min, max)
+   return function(str, k)
+      local str = assert(str, 'missing value for '..k)
+      local start = 1
+      local is_negative
+      local base = 10
+      if str:match('^-') then start, is_negative = 2, true
+      elseif str:match('^+') then start = 2 end
+      if str:match('^0x', start) then base, start = 16, start + 2
+      elseif str:match('^0', start) then base = 8 end
```
Comment (kbara) [inline src/lib/yang/data.lua]: Are we required to support octal? (Both supporting and not supporting it tend to be rich sources of misconfiguration.)

Comment (wingo) [reply src/lib/yang/data.lua]: yes we are required to support it in the sense that the syntax for yang integers in the XML mapping, which we are using as a model for our syntactic mapping, supports octal.

**Comment by wingo** [comment]
A bit of a design doc here, from a README.yang.md that I am writing:

### Snabb program configuration with YANG (lib.yang)

YANG is a data modelling language designed for use in networking
equipment, standardized as [RFC
6020](https://tools.ietf.org/html/rfc6020).  The `lib.yang` modules
provide YANG facilities to Snabb applications, allowing operators to
understand how to work with a Snabb data plane and also providing
convenient configuration facilities for data-plane authors.

Everything in YANG starts with a _schema_: a specification of the data
model of a device.  For example, consider a simple Snabb router that
receives IPv4 traffic and sends it out one of 12 ports.  We might
model it like this:

``` yang
module snabb-simple-router {
  namespace snabb:simple-router;
  prefix simple-router;

  import ietf-inet-types {prefix inet;}

  leaf active { type boolean; default true; }

  container routes {
    presence true;
    list route {
      key addr;
      leaf addr { type inet:ipv4-address; mandatory true; }
      leaf port { type uint8 { range 0..11; }; mandatory true; }
    }
  }
}
```

Given this schema, `lib.yang` can automatically derive a configuration
file format for this Snabb program and create a parser that applies
the validation constraints from the schema.  The result is a simple
plain-old-data Lua object that the data-plane can use directly.

Additionally there is support for efficient binary compilation of
configurations.  The problem is that even in this simple router, the
routing table can grow quite large.  While particular applications can
sometimes incrementally update their configurations without completely
reloading the configuration from the start, in general reloading is
almost always a possibility, and you want to avoid packet loss during
the time that the millions of routing table entries are loaded and
validated.

For that reason the `lib.yang` code also defines a mapping that, given
a YANG schema, can compile any configuration for that schema into a
pre-validated binary file that the data-plane can just load up
directly.  Additionally for `list` nodes that map between keys and
values, the `lib.yang` facilities can compile that map into an
efficient [`ctable`](../README.ctable.md), letting the data-plane use
the configuration as-is.

**Comment by wingo** [comment]
additional details about how to use that example schema and how it should map to configurations forthcoming

**Comment by wingo** [comment]
PTAL @xray7224, I think this is in a good state to merge to Igalia/yang.  See README.md for docs.

File: src/lib/yang/README.md [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+Inside the `routes` container is the `route` list, which is also
+represented as a table.  Recall that in YANG, `list` types are really
+key-value associations, so the `route` table has a `:lookup` method to
+get its sub-items.  Therefore to get the port for address 1.2.3.4, you
+would do:
+
+```lua
+local yang = require('lib.yang.yang')
+local ipv4 = require('lib.protocol.ipv4)
+local data = yang.load_data_for_schema(router_schema, conf_str)
```
Comment (tsyesika) [inline src/lib/yang/README.md:237]: You're missing a quote.

File: src/lib/yang/README.md [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+they are positive, or `int64_t` otherwise.
+
+Boolean values are represented using normal Lua booleans, of course.
+
+String values are just parsed to Lua strings, with the normal Lua
+limitation that UTF-8 data is not decoded.  Lua strings look like
+strings but really they are byte arrays.
+
+There is special support for the `ipv4-address`, `ipv4-prefix`,
+`ipv6-address`, and `ipv6-address` types from `ietf-inet-types`, and
```
Comment (tsyesika) [inline src/lib/yang/README.md]: You have "ipv6-address" twice, I think you mean "ipv6-prefix"

File: src/lib/yang/data.lua [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+-- FIXME :)
+local function range_validator(range, f) return f end
+local function length_validator(range, f) return f end
+local function pattern_validator(range, f) return f end
+local function bit_validator(range, f) return f end
+local function enum_validator(range, f) return f end
+
+local function value_parser(typ)
+   local prim = typ.primitive_type
+   local parse = assert(value.types[prim], prim).parse
```
Comment (tsyesika) [inline src/lib/yang/data.lua:68]: This error message could be more useful, currently it's just "Error in core/main.lua:26: enumeration".

**Review by tsyesika — COMMENTED** [review]
Everything looks pretty good on the code side, just a few little things I noticed in the documentation and the error message.

I noticed the lack of a few types, some of which are documented. One of these is the `enumeration`  which is used in our own yang lwaftr document (for PacketPolicy).

Finally our yang document (snabb-softwire.yang) needs a few updates in addition issues around presence. Those are an import of `ietf-yang-types` and modifying where mac-address comes from, otherwise your code rightly rejects it.

**Comment by wingo** [comment]
Addressed nits, merging.  Thanks again for the review :)

**Pull request merged**